core/types/calls.cc

```diff
@@ -1363,6 +1363,31 @@ class Magic_buildArray : public IntrinsicMethod {
     }
 } Magic_buildArray;
 
+class Magic_buildRange : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
+        ENFORCE(args.args.size() == 2, "Magic_buildRange called without argument");
+
+        auto rangeElemType = Types::dropLiteral(args.args[0]->type);
+        auto firstArgIsNil = rangeElemType->isNilClass();
+        if (!firstArgIsNil) {
+            rangeElemType = Types::dropNil(gs, rangeElemType);
+        }
+        auto other = Types::dropLiteral(args.args[1]->type);
+        auto secondArgIsNil = other->isNilClass();
+        if (firstArgIsNil) {
+            if (secondArgIsNil) {
+                rangeElemType = Types::untypedUntracked();
+            } else {
+                rangeElemType = Types::dropNil(gs, other);
+            }
+        } else if (other->isUntyped() || (!secondArgIsNil && !Types::equiv(gs, other, rangeElemType))) {
```

Comment:
```suggestion
        } else if (!secondArgIsNil) {
```

I think you can simplify the conditional here quite a bit.